[Step3]자동차 경주

re-mora opened the pull request:
안녕하세요 리뷰어님, 교육생 최용석입니다 :)

3단계 `자동차 경주` 미션 구현이 완료되어 리뷰 요청 드립니다.

---

저번 미션에서 주신 피드백을 정리하여 아래 커밋에 적용하였습니다.
- [step2 문자열 계산기 코드 리뷰 피드백 내용 정리](https://github.com/next-step/kotlin-racingcar/commit/098647bc8ba6019e9ecbabb1d8fe6c5ccf870fd6)
- [일급 시민을 활용한 연산자 enum 객체 리팩토링(OperationType)](https://github.com/next-step/kotlin-racingcar/commit/082580db70f382a7d0df608dcba03eb84fd4bbd0)
- [계산기 객체 재활용을 위한 리팩토링(StringCalculator)](https://github.com/next-step/kotlin-racingcar/commit/1b44b3cf05304c5e961972a725371eca7d399234)
- [step2 문자열 계산기 구현 회고](https://github.com/next-step/kotlin-racingcar/commit/aa0d8fa1fe37856f98a8eb8ffaeac4f536341a9e)

---
이번 미션은 아래와 같은 사항에 중점을 두어 구현하였습니다.

### package 구조 
- domain : 자동차 경주를 진행 시 각 상태를 표현하는 객체 모임
  - Car : 경주에 참여하는 자동차 객체
  - Cars : 경주에 참여하는 자동차 목록 일급 컬렉션
  - RoundResult : 각 라운드의 경주 결과를 표현하는 객체
  - RoundResults : 각 차량의 전체 경주 결과를 표현하는 일급 컬렉션
  - MoveStatus : 주어진 숫자에 따라 전진/멈춤을 판별하여 상태를 반환하는 enum
- expansion: common utils 성격으로 사용될 수 있는 함수
- service : 자동차 경주 진행 부
- controller : 입력 받은 값을 서비스 계층에 전달, 게임 결과를 반환받아 출력부에 전달하는 역할
- view : 자동차 경주 진행에 필요한 입출력 처리부

### test 작성
- kotest BehaviorSpec을 이용한 TC 작성

---

접근이 잘못되었거나, 부족한 부분이 있다면 이번 미션 역시 가감없이 피드백 부탁드립니다. 😄

re-mora committed:
docs: 문자열 계산기 요구사항 및 TODO List 작성

re-mora committed:
feat: 문자열로 입력된 연산자에 따른 타입 추출 및 수식 계산 부 구현 및 TC 작성

- Java BiFunction 인터페이스와 유사한 방식으로 연산자 타입에 따른 수식 계산 부를 enum으로 구현
- 기능 구현에 따른 문서 갱신

re-mora committed:
feat: 연산자가 여러개인 표현식의 사칙연산 계산기 구현 및 TC 작성

- 연산자가 한개인 표현식의 사칙연산 계산 TC 작성
- 연산자가 여러개인 표현식의 사칙연산 계산 TC 작성
- 기능 구현에 따른 문서 갱신

re-mora committed:
feat: 복잡한 수식 TC 작성을 통한 예외 처리 보강

예외 처리 추가
- 정의된 연산자 타입이 아닌 경우 예외 처리 및 TC 작성
- 제수가 0인 경우 예외 처리 및 TC 작성
- 입력값이 유효하지 않은 경우 예외 처리
  - 계산 항목이 부족한 경우
  - 계산 항목이 숫자가 아닌 경우
- Double 타입으로 연산한 결과가 정수인 경우 소수점 영역 제외 처리

re-mora committed:
test: 계산 항목의 숫자 여부 판별 부 수정 및 음수와 실수가 포함된 사칙연산 TC 추가 작성

re-mora committed:
fix: 깨지는 TC 수정

re-mora committed:
type: 잘못 표현된 TC 설명 수정

re-mora committed:
docs: 계산기 관련 용어 정리 작성

re-mora committed:
test: JUnit5 기반으로 작성된 테스트를 kotest로 작성

- JUnit5의 @ParameterizedTest를 대체하는 kotest의 forAll()을 이용한 여러 파라미터에 대한 테스트 작성
- JUnit5의 assertThatExceptionOfType()을 대체하는 kotest shouldThrow()를 이용한 예외 발생 테스트 작성

re-mora committed:
typo: 항을 표현하는 변수명의 올바르지 못한 표기 수정

- AS-IS : leftValue, rightValue
- TO-BE : leftTerm, rightTerm

re-mora committed:
type: 잘못 표기된 package 이름 변경

AS-IS : step1
TO-BE : step2

re-mora committed:
docs: 3단계 자동차 경주 요구사항 및 구현 기능 목록 작성

re-mora committed:
docs: step2 문자열 계산기 코드 리뷰 피드백 내용 정리

re-mora committed:
refactor: step2 코드 리뷰 피드백 적용(OperationType)

- 일급 시민을 이용하여 Enum의 각 상태에 해당하는 함수를 구현하기 위해 부가적으로 작성된 코드 제거
- Kotlin의 requrie(), requireNotNull() 적용을 통해 유효성 검증 부의 구현 편의성 및 가독성 개선
- 약어로 표현된 변수의 구체적인 의미 표현을 통한 가독성 개선
- Double 타입 파라미터의 유효성 검증을 위한 확장 함수 작성
- String.format()으로 작성된 코드를 String의 확장 함수 format()으로 대체

re-mora committed:
refactor: step2 코드 리뷰 피드백 적용(StringCalculator)

re-mora committed:
docs: step2 문자열 계산기 구현 회고

re-mora committed:
docs: step3 자동차 경주 구현 기능 목록 수정

re-mora committed:
feat: 주어진 숫자에 따라 전진/멈춤 상태를 결정하는 MoveStatus 구현

- 주어진 숫자가 4이상인 경우, 전진 상태 반환
- 주어진 숫자가 4미만인 경우, 멈춤 상태 반환
- 주어진 숫자의 4이상 여부를 판별하기 위한 Int 확장함수 작성
- 기능 구현에 따른 문서 갱신

re-mora committed:
feat: 자동차 경주를 진행하고 결과를 반환하는 자동차 객체 설계

- 자동차 객체 속성 설계
  - id : 경주에 참여하는 각 자동차를 식별하기 위한 필드
  - mveStatues : 각 랑룬드의 전진/몸춤 결과 목록을 표현하는 일급 컬렉션
- 단일 차량의 단일 라운드 진행 및 결과 검증 TC 작성
- 단일 차량의 여러 라운드 진행 및 결과 검증 TC 작성

re-mora committed:
feat: 제한된 범위의 난수 발생기 구현

re-mora committed:
feat: 자동차 경주 시도 횟수와 경주에 참여할 자동차 수 입력 부

re-mora committed:
feat: 자동차 경주에 참여하는 자동차 목록을 표현하는 일급 컬렉션 설계

- 참여할 자동차 수 만큼 자동차 객체를 생성하여 일급컬렉션으로 반환하는 생성기 구현

re-mora committed:
feat: 자동차 경주 진행 부 구현

- 자동차 경주 시작을 위한 main 함수 작성
- 자동차 경주에 필요한 값 입력 Controller 작성
- 자동차 경주 진행 Service 작성
- 각 라운드 완료 시 결과 출력 부 작성

re-mora committed:
refactor: for문 작성 시 생략 가능한 step 1 구문 삭제

re-mora committed:
fix: lint 깨짐 수정

re-mora committed:
docs: 자동차 경주 진행부와 출력부가 결합된 부분 개선을 위한 TODO List 추가

re-mora committed:
docs: 유효하지 못한 값을 입력하는 경우, 값 입력 재시도 처리 TODO List 추가

re-mora committed:
refactor: 각 라운드 결과를 반환하는 RoundResult 및 누적된 전체 라운드 결과를 반환하는 RoundResults 일급 컬렉션 설계

- 각 라운드의 경주 결과 반환 검증 TC 작성
- 누적된 전체 라운드 경주 결과 반환 검증 TC 작성

re-mora committed:
refactor: 자동차 객체에 전체 라운드의 결과 목록 일급 컬렉션 필드 적용

re-mora committed:
refactor: 자동차 경주 로직과 출력부가 결합된 부분 분리

- 유효하지 못한 값을 입력하는 경우, 값 입력 재시도 처리

re-mora committed:
Merge branch 'mission/step/#3-racingcar' into mission/step/#3-racingcar-solve-conflict

# Conflicts:
#	docs/STEP2-README.md
#	src/main/kotlin/step2/calculator/OperationType.kt
#	src/main/kotlin/step2/calculator/StringCalculator.kt
#	src/test/kotlin/step2/calculator/OperationTypeKoTest.kt
#	src/test/kotlin/step2/calculator/OperationTypeTest.kt
#	src/test/kotlin/step2/calculator/StringCalculatorKoTest.kt
#	src/test/kotlin/step2/calculator/StringCalculatorTest.kt

malibinYun commented on src/main/kotlin/step2/calculator/Expansion.kt:
```diff
@@ -0,0 +1,6 @@
+fun Double.isNotZero(): Boolean = this != 0.0
+fun Double.hasNoRemainder(): Boolean = this % 1 == 0.0
+
+fun Int.overThan(target: Int): Boolean = this >= target
```
isNotZero와 overThan은 기호로 나타내는 것이 개인적으로 가독성이 좀 더 나아 보이네요 :)

여담으로, 코틀린에는 class에 compareTo 연산자를 오버라이드 하면, `<`, `>`, `<=`, `>=` 기호를 사용할 수 있게 된답니다!
```kotlin
class A {
     operator fun compareTo(old: A): Int {
        //...
    }
}

val a1 = A()
val a2 = A()
if (a1 > a2) {
    // ...
}
```

malibinYun commented on src/main/kotlin/step2/calculator/ExpressionParser.kt:
```diff
@@ -0,0 +1,17 @@
+    private fun validateExpression(expression: String) {
+        val expressionLength: Int = expression.trim().length
+        require(expressionLength.overThan(MINIMUM_EXPRESSION_LENGTH)) { LESS_THAN_MINIMUM_EXPRESSION_LENGTH_ERROR_MESSAGE }
+    }
```
Exception Message에 어떤 수식이 들어왔는지도 찍어주는 것은 어떨까요?

malibinYun commented on src/main/kotlin/step2/calculator/OperationType.kt:
```diff
@@ -1,50 +1,30 @@
+        fun match(operator: String): OperationType =
+            requireNotNull(
+                values().firstOrNull { it.operator == operator }
+            ) { INVALID_OPERATOR_ERROR_MESSAGE.format(operator) }
```
elvis operator를 활용해볼 수도 있겠네요 :)
```suggestion
        fun match(operator: String): OperationType = values().find{ it.operator == operator }
            ?: throw IllegalArgumentException(INVALID_OPERATOR_ERROR_MESSAGE.format(operator))
```

malibinYun commented on src/main/kotlin/step3/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,19 @@
+
+class Car {
+    val id: String = UUID.randomUUID().toString().substringBefore(CAR_ID_DELIMITER)
+    var roundResults: RoundResults = RoundResults()
+
+    fun race(round: Int, randomNumber: Int): RoundResult =
+        RoundResult(round, MoveStatus.match(randomNumber)).apply {
+            roundResults.accumulate(this)
+        }
```
자동차가 각 라운드의 결과를 저장하는 역할까지 가지지 않아도 된다고 생각해요.
자동차는 거리만 가지고 있고, 각 라운드 결과는 자동차 외부에서 관리해보는 것은 어떨까요?

malibinYun commented on src/main/kotlin/step3/racingcar/domain/Cars.kt:
```diff
@@ -0,0 +1,5 @@
+package step3.racingcar.domain
+
+class Cars(val elements: MutableList<Car> = mutableListOf()) {
+    fun size() = elements.size
+}
```
Cars가 개수를 반환하는 것 말고는 아무 역할을 하지 않아 불필요해 보입니다.

추가로, elements가 MutableList로 공개되어 있어서, 외부에서 해당 변수를 접근해서 내부 변경이 가능합니다.
그대로 Cars를 활용하게 된다면, 캡슐화 해서 외부로부터 변경을 막아야겠네요 :)

malibinYun commented on src/main/kotlin/step3/racingcar/domain/RoundResults.kt:
```diff
@@ -0,0 +1,15 @@
+package step3.racingcar.domain
+
+class RoundResults {
+    private var elements: MutableList<RoundResult> = mutableListOf()
+
```
elements 자체를 재할당하는 경우는 없어서, val로 변경해보는 것은 어떨까요?

malibinYun commented on src/main/kotlin/step3/racingcar/domain/RoundResults.kt:
```diff
@@ -0,0 +1,15 @@
+
+    fun size(): Int = elements.size
+
+    fun elements() = elements
+}
```
이 경우 MutableList의 참조를 그대로 밖으로 노출시킵니다.
외부에서 이 메서드로부터 받은 MutableList에 add 혹은 remove 등의 메서드를 호출하면 어떻게 될까요?

malibinYun commented on src/main/kotlin/step3/racingcar/service/RacingCarService.kt:
```diff
@@ -0,0 +1,23 @@
+    fun play(carCount: Int, roundCount: Int): Cars {
+        val cars: Cars = CarGenerator.generate(carCount)
+
+        for (i in 1..roundCount) {
+            playEachRound(i, cars)
+        }
```
repeat() 메서드를 활용해보는 것은 어떨까요?

malibinYun commented on src/main/kotlin/step3/racingcar/utils/CarGenerator.kt:
```diff
@@ -0,0 +1,14 @@
+
+object CarGenerator {
+    fun generate(carCount: Int): Cars = Cars(
+        mutableListOf<Car>().also {
+            for (i in 0 until carCount) {
+                it.add(Car())
+            }
+        }
```
_List_(...) { ... } 전역 함수를 활용해보는 것은 어떨까요?

malibinYun commented on src/test/kotlin/step3/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,41 @@
+import io.kotest.core.spec.style.BehaviorSpec
+import io.kotest.matchers.shouldBe
+
+class CarTest : BehaviorSpec({
+    given("Car : 단일 차량의 여러 라운드 진행") {
+        var round = 0
+        val car = Car()
```
한 번에 시나리오 테스트를 해보는 것도 좋지만, 먼저 아래 항목들을 명확하게 테스트해보는 것은 어떨까요?
* 자동차가 움직이는 경우
* 자동차가 움직이지 않는 경우

malibinYun reviewed (changes requested):
3단계 미션 고생 많으셨어요!!
상세한 요구사항을 적어주시고, 그에 따라 세밀하게 커밋을 해주신 게 정말 인상 깊네요 👍👍
정말 많은 생각을 하시면서 미션을 진행하셨다는 게 느껴지네요 🔥
생각해보면 좋을 점들에 코멘트 남겨두었어요 :)
피드백 반영하시구 다시 리뷰 요청 부탁 드릴게요!

re-mora committed:
refactor: step2 2차 코드리뷰 피드백 적용

- 확장 함수로 표현한 관계 연산자(<,<=,>,>=)를 직접 명시하여 가독성 개선
- 예외가 발생한 요청값 확인이 용이하도록 예외 메세지에 예외 발생 파라미터 명시
- kotlin elvis operator를 이용하여 객체값이 null인 경우 기본 동작 처리 정의

re-mora committed:
docs: step3 자동차 경주 코드 리뷰 피드백 내용 정리

re-mora committed:
refactor: 역할이 명확하지 않은 Cars 객체에 자동차 경주에 참여하는 자동차 목록을 생성하는 책임 부여

re-mora committed:
refactor: 가변 타입으로 선언된 필드 반환 시, 외부에서의 변경을 제한하기 위해 불변객체로 반환 처리

re-mora committed:
docs: 자동차 경주 구현 기능 목록 수정

경기 진행 구현 방식 변경
- AS-IS : 각 자동차의 라운드 결과를 별도의 응답 객체에 누적 후 경기 결과를 출력
- TO-BE : 각 자동차에 거리를 표현하는 필드에 진행 정도를 누적 후 경기 결과를 출력

re-mora committed:
refactor: 경기 진행 구현 방식 변경

- AS-IS : 각 자동차의 라운드 결과를 별도의 응답 객체에 누적 후 경기 결과를 출력
- TO-BE : 각 자동차에 거리를 표현하는 필드에 진행 정도를 누적 후 경기 결과를 출력

re-mora committed:
refactor: repeat()를 이용하여 for문으로 작성된 반복문의 가독성 개선

re-mora committed:
refactor: Service에서 객체를 생성하는 책임을 Controller로 분리

re-mora committed:
refactor: 미사용 코드 제거 및 코드 정리

re-mora commented:
안녕하세요 리뷰어님!, 교육생 최용석입니다 :)

기존 각 차량의 매 라운드 경기 결과를 일급컬렉션에 누적시킨 후 반환 받아 별도의 출력 부에서 경기 결과를 출력 하도록
구현한 부분의 복잡도가 높아져, 경주에 참가하는 `Car`객체에 주행 거리 필드를 추가하고, 매 경기 마다 주행거리 필드에 전진/멈춤 결과를 누적하여 복잡도를 떨어트려 보았습니다.

- [경기 진행 구현 방식 변경](https://github.com/next-step/kotlin-racingcar/pull/813/commits/c607d8caf84fa4d07bbb7879ab74f173188bc8b4)

또한 리뷰 주신 내용을 아래와 같이 정리한 후 수정하였습니다.
- [refactor: step2 2차 코드리뷰 피드백 적용](https://github.com/next-step/kotlin-racingcar/pull/813/commits/3ddd056f6cab0479b9c75ef035f9bf3301e83de1)

- [docs: step3 자동차 경주 코드 리뷰 피드백 내용 정리](https://github.com/next-step/kotlin-racingcar/pull/813/commits/c0941bdb6548f4b4a11905f856455dac4f5836f5)

미흡한 부분이 있다면 한번 더 리뷰 부탁드립니다!

---

날씨가 많이 추워졌는데, 항상 감기 조심하시고, 즐거운 주말 보내세요! 😄

malibinYun commented on src/main/kotlin/step2/calculator/Expansion.kt:
```diff
@@ -0,0 +1,3 @@
+package step2.calculator
```
파일 명을 늦게 확인했네요!
`확장 함수`의 영어 명칭은 `Extension Function` 입니다 :)

malibinYun commented on src/main/kotlin/step3/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,18 @@
+        }
+    }
+
+    private fun isMove(randomNumber: Int) = randomNumber >= 4
```
4 같은 매직 넘버를 상수로 분리해보는 것은 어떨까요?
이름을 부여함으로써 의미 부여도 할 수 있겠네요 :)

malibinYun commented on src/main/kotlin/step3/racingcar/domain/Cars.kt:
```diff
@@ -0,0 +1,7 @@
+
+class Cars(count: Int) {
+    private val elements: List<Car> = List(count) { Car() }
+    fun elements() = elements
+    fun size() = elements.size
```
https://github.com/next-step/kotlin-racingcar/pull/739#discussion_r1014619108
이 코멘트를 다시 한 번 참고해주세요 :)

malibinYun commented on src/main/kotlin/step3/racingcar/domain/PlayInfo.kt:
```diff
@@ -0,0 +1,5 @@
+package step3.racingcar.domain
+
+class PlayInfo(numberOfCars: Int, val totalRound: Int) {
+    val cars: Cars = Cars(numberOfCars)
+}
```
자동차 경주에 필요한 정보가 Cars를 만들어내는 역할을 가지는 것은 어색해보여요.
정보들을 전달만 하고, Cars는 외부에서 만드는 것은 어떨까요?

malibinYun commented on src/main/kotlin/step3/racingcar/service/RacingCarService.kt:
```diff
@@ -0,0 +1,22 @@
+
+    private fun playEachRound(currentRound: Int, cars: Cars) {
+        cars.elements().forEach {
+            val randomValue = RandomNumberGenerator.generate()
```
랜덤 숫자를 생성해 주는 것은 Service 밖에서 주입받아보는 것은 어떨까요?
수업에서 배운 내용처럼, Service도 테스트가 가능해지겠네요 :)

malibinYun commented on src/test/kotlin/step3/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,50 @@
+import io.kotest.matchers.shouldBe
+
+class CarTest : BehaviorSpec({
+    given("경주에 참가하는 자동차 한대에 숫자 4가 주어지고") {
+        val givenCar = Car()
+        val givenNumber = 4
+        `when`("경주를 진행하면") {
+            givenCar.race(givenNumber)
+            then("현재 차량의 주행거리를 1만큼 누적한다.") {
+                givenCar.distance shouldBe 1
+            }
+        }
+    }
+
+    given("경주에 참가하는 자동차 한대에 숫자 3가 주어지고") {
+        val givenCar = Car()
+        val givenNumber = 3
+        `when`("경주를 진행하면") {
+            givenCar.race(givenNumber)
+            then("현재 차량의 주행거리는 누적되지 않는다.") {
+                givenCar.distance shouldBe 0
+            }
+        }
+    }
```
4이상, 3이하 등으로 표현하는 것은 어떨까요?
테스트를 처음 보는 사람이 3, 4 이외의 숫자는 어떻게 되는 지 혼란스러울지도 모르겠네요 :)

malibinYun reviewed (approved):
3단계 미션 고생 많으셨습니다 :)
커밋도 상세히 나눠주시고, 리뷰 받은 부분들을 정리해보시기도 하시다니..
정말 정성이 느껴지는 미션이네요 👍
생각해보면 좋을 점들에 코멘트 남겨두었어요.
다음 미션을 위해 머지하겠습니다. 피드백은 다음 미션에 반영해주세요!
용석님도 감기 조심하시구, 즐거운 미션 주말(?) 보내세요! 😆
